text/3476-impl-trait-for-fn.md

```diff
@@ -0,0 +1,90 @@
+- Feature Name: `impl-trait-for-fn`
+- Start Date: 2023-08-20
+- RFC PR: [rust-lang/rfcs#3476](https://github.com/rust-lang/rfcs/pull/3476)
+- Rust Issue: N/A
+
+# Summary
+[summary]: #summary
+
+Support for implementing traits on functions
+
+# Motivation
+[motivation]: #motivation
+
+I was trying to make something similar to bevy's system functions. And for safety reasons, they check for conflicts between SystemParams, so that a function requiring `Res<A>` and `ResMut<A>` [panic](https://github.com/bevyengine/bevy/blob/main/crates/bevy_ecs/src/system/system_param.rs#L421).
+
+Then after I heard about axum's [`#[debug_handler]`](https://docs.rs/axum/latest/axum/attr.debug_handler.html) I wanted to do something similar to my copy of bevy systems, so that I get compile time errors when there is a conflict. I wanted even more, I wanted to force the user to mark the function with a specific proc attribute macro in order to make it possible to pass it into my code and call itself a system.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As we all know, you can implement a trait for a struct with the following syntax
+```rust
+struct Timmy;
+impl Person for Timmy {
+    fn greet() {
+        println!("Hey it's me, Timmy!");
+    }
+}
+```
+And we can also implement a trait for all functions with a specific signature like this
+```rust
+impl<F> ValidSignature for F
+    where F: Fn(i32) -> bool
+{
+    /* ... */
+}
+```
+Now we can also implement traits for specific functions only
+```rust
+fn valid() {}
+impl ValidFunction for fn valid {
+    /* ... */
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+It gives the possibility to also implement a trait directly scoped to a function instead of generic implementation of multiple. Other than that, it basically behaves the same. It should also be possible to implement them via proc attribute macros:
+```rust
+#[impl_debug_name = "Greeting"]
+fn greet() {
+    /* ... */
+}
+// should expand to something like
+fn greet() {
+    /* ... */
+}
+impl FnDebugName for fn greet() {
+    fn debug_name() -> &'static str {
+        "Greeting"
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+i dont know any
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+I think it's a easy task because we can already implement traits for a group of functions with the same signature, so why shouldn't we also implement single scoped impls?
+
+# Prior art
+[prior-art]: #prior-art
+
+i dont know any
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Is the syntax good? I feel like we could drop the `fn` to from `impl Trait for fn function` to `impl Trait for function`.
```

Comment:
This makes sense, I think - lints will make sure types are styled properly, and the `impl Trait for function` syntax already works to differentiate between a function and type due to it obviously being a function, at least in the reference you gave.

One question - what about input and output types for the function? You gave an example of a function that has no input and no output types. Would it be like how other types work now, with it just being `impl Trait for any_function()` and `fn any_function(x: i32) -> 32`?

Reply:
I've on purpose left out the params parens and return type arrows in the impl so that you don't have to write them.
So
```rust
fn itoa(i32) -> String { .. }
```
would still be implemented with
```rust
impl MyTrait for fn itoa { .. }
```
This is possible because you cannot have two functions with different params but the same name. Im also quite sure that we also have no complications with targeting what fn/type to implement, but if so, there's the `fn` keyword in the impl. Otherwise we could think about dropping it.

But yeah this is was not well written by me. I'll change a fn signature in an example so it's easier to understand.

Reply:
I'd say it's fine to get rid of the `fn` keyword, but only so long as the function name is *clearly* a function in the `impl` bit, like by having any function, no matter its input types or outputs, get traits with `impl Trait for function()`

Reply:
@orowith2os can this be marked as resolved?

Reply:
😮 now i see where you got that name() idea from. I mistyped in one example 😅

Reply:
So long as there's either the `fn` keyword in the `impl` or some identifier (`()`) to make it clear that it's a function, lgtm :)